File: 그리디/leetcode409(Easy)-Longest Palindrome.js
// 65ms(75%), 44MB(26%)
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindrome = function (s) {
  const counter = Array.from(s).reduce((obj, char) => {
    if (!obj[char]) obj[char] = 0
    obj[char] += 1
    return obj
  }, {})

  let ans = 0
  for (const [key, value] of Object.entries(counter)) {
    if (value % 2 === 0) {
      ans += value
      counter[key] = 0
    } else {
      if (value > 2) {
        ans += value - 1
        counter[key] = 1
      }
    }
  }

  for (const value of Object.values(counter)) {
    if (value === 1) {
      return ans + 1
    }
  }
  return ans
}

File: 그리디/leetcode409(Easy)-Longest Palindrome.py
# 29ms(90%), 13MB(53%)
class Solution:
    def longestPalindrome(self, s: str) -> int:
        counter = collections.Counter(list(s))

        ans = 0
        # 일단 짝수개는 계속 더해줌
        for key, value in counter.items():
            # 짝수개라면 모두 더해주고, 남은 개수를 0으로 만듬
            if value % 2 == 0:
                ans += value
                counter[key] = 0
            else:
                # 3이상의 홀수개라면 
                if value > 2:
                    ans += value - 1
                    counter[key] = 1
        
        # 남은 개수가 1인 것중 하나만 더해줌
        for key, value in counter.items():
            if ans % 2 == 1:
                break
            if value == 1:
                ans += 1

        return ans